tkt-69267: Add --debug(-d) and --images(-i) flag to clean

skarekrow opened the pull request:
Also switch over to subprocess zfs as occasionally the base dataset would not be destroyed properly by py-libzfs.

FreeNAS Ticket: #69267

Commit:
Add --debug(-d) and --images(-i) flag to clean

Also switch over to subprocess zfs as occasionally the base dataset would not be destroyed properly by py-libzfs.

FreeNAS Ticket: #69267

Commit:
Add new property to template created jails

- This allows us to destroy jails faster that are from templates.
- Destroy leftovers properly now
- Add iocroot_datasets property
- Remove context manager for zfs_get_property as it's not functioning, Waqar will have a PR for this and the other usages.

Commit:
Clarify message on debug cleaning

Commit:
Fix clean breaking from destroy behavior change

Commit:
Remove debug

Commit:
Fix thickjails key

Commit:
Use stdout, stderr instead of capture_output for python3.6

sonicaj commented on iocage_lib/ioc_create.py:
```diff
@@ -628,6 +633,9 @@ def create_config(self, jail_uuid, release):
             jail_props.update(d_conf)
             jail_props['CONFIG_TYPE'] = 'THICK'
 
+        if source_template is not None:
```
Should we perhaps take care of this prop as well as we discussed ?

skarekrow replied:
That's a different PR to rename that, unless you're talking about something else?

sonicaj replied:
Yup, I'm referring to renaming that. But if we don't do that, we would not be cleaning up properly, thoughts ?

sonicaj replied:
Oh, I think I misunderstood you. Are you saying we should take care of that use case in another PR ?

skarekrow replied:
Yep :)

sonicaj commented on iocage_lib/ioc_destroy.py:
```diff
@@ -79,147 +79,138 @@ def __stop_jails__(datasets, path=None, root=False, clean=False):
-            # It gives us a string that says "none", not terribly
-            # useful, fixing that.
-            _path = _path if _path != "none" else None
+            _path = mountpoint.replace('/root', '')
```
Perhaps we think more about this as well ?

sonicaj commented on iocage_lib/ioc_destroy.py:
```diff
@@ -79,147 +79,138 @@ def __stop_jails__(datasets, path=None, root=False, clean=False):
-
-        try:
-            path = dataset.properties["mountpoint"].value
+        uuid = dataset.rsplit('/root')[0].split('/')[-1]
```
This would be related to the above one as well
`'vol1/iocage/jails/root/root'.rsplit('/root')[0].split('/')[-1]`

sonicaj commented on iocage_lib/ioc_json.py:
```diff
@@ -303,6 +322,53 @@ def zfs_get_snapshot(self, snap_id):
+                               f'{identifier} failed!\n'
+                               f'Reason: {e.stderr.decode()}'
+                },
+                _callback=self.callback,
```
Would it be possible that this raises an undefined error if somewhere we use this function and callback isn't defined ?

skarekrow replied:
callback will be None in that case, which is expected most of the time

sonicaj replied:
So if we use IOCZFS directly, it would not exist then

sonicaj commented on tests/functional_tests/0013_import_test.py:
```diff
@@ -61,10 +61,3 @@ def test_01_import_jail(invoke_cli, jail, skip_test, remove_file, zfs):
 
     assert jail.exists is True, f'{exported_jail} jail did not import'
-
-    # Let's do a clean up on the exported files
```
I think we should let these be in case we are running selected tests and then import would not clean up on the images dataset ? Thoughts ?

skarekrow replied:
I suppose we can, but we don't officially support running selected tests, right now it's the whole battery

Commit:
Handle jails named 'root' and fix a string typo

Commit:
More fixes

- Use ioc_exceptions to suppress
- Properly handle jails named 'root' for templates
- Destroy leftover parent datasets of jails when a template is destroyed
- Reverse dataset direction to destroy in the correct order
- Add mountpoint special handling to zfs_get_property
- Use zfs_get_property for the hack88 portion to avoid superfluous messages about non-existent datasets during cleaning.

Commit:
Travis

Commit:
Add callback to IOCZFS

Commit:
Travis
